fpga_interchange/dedicated_interconnect.cc

```diff
@@ -0,0 +1,785 @@
+            // driver is unplaced.  Once the driver is placed, this constraint
+            // can be evaluated.
+            if (ctx->debug) {
+                log_info("BEL %s is not valid because driver cell %s/%s is not placed\n", ctx->nameOfBel(bel),
+                         net->driver.cell->name.c_str(ctx), net->driver.port.c_str(ctx));
+            }
+            return false;
+        }
+
+        const auto &driver_bel_data = bel_info(ctx->chip_info, driver_bel);
+
+        if (bel.tile == driver_bel.tile && bel_data.site == driver_bel_data.site) {
+            // This is a site local routing, even though this is a sink
+            // with a dedicated interconnect.
+            continue;
+        }
+
+        Loc driver_loc = ctx->getBelLocation(driver_bel);
+
+        DeltaTileTypeBelPin driver_type_bel_pin;
+        driver_type_bel_pin.delta_x = driver_loc.x - bel_loc.x;
+        driver_type_bel_pin.delta_y = driver_loc.y - bel_loc.y;
+        driver_type_bel_pin.type_bel_pin.tile_type = ctx->chip_info->tiles[driver_bel.tile].type;
+        driver_type_bel_pin.type_bel_pin.bel_index = driver_bel.index;
+        driver_type_bel_pin.type_bel_pin.bel_pin =
+                get_only_value(ctx->getBelPinsForCellPin(net->driver.cell, net->driver.port));
+
+        // Do fast routing check to see if the pair of driver and sink
+        // every are valid.
+        if (iter->second.count(driver_type_bel_pin) == 0) {
+            if (ctx->debug) {
+                log_info("BEL %s is not valid because pin %s cannot be driven by %s/%s\n", ctx->nameOfBel(bel),
+                         bel_pin.c_str(ctx), ctx->nameOfBel(driver_bel),
+                         driver_type_bel_pin.type_bel_pin.bel_pin.c_str(ctx));
+            }
+            return false;
+        }
+
+        // Do detailed routing check to ensure driver can reach sink.
+        //
+        // FIXME: This might be too slow, but it handles a case on
+        // SLICEL.COUT -> SLICEL.CIN has delta_y = {1, 2}, but the
+        // delta_y=2 case is rare.
+        if (!check_routing(driver_bel, driver_type_bel_pin.type_bel_pin.bel_pin, bel, bel_pin)) {
+            if (ctx->debug) {
+                log_info("BEL %s is not valid because pin %s cannot be driven by %s/%s (via detailed check)\n",
+                         ctx->nameOfBel(bel), bel_pin.c_str(ctx), ctx->nameOfBel(driver_bel),
+                         driver_type_bel_pin.type_bel_pin.bel_pin.c_str(ctx));
+            }
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool DedicatedInterconnect::isBelLocationValid(BelId bel, const CellInfo *cell) const
+{
+    NPNR_ASSERT(bel != BelId());
+
+    for (const auto &port_pair : cell->ports) {
+        IdString port_name = port_pair.first;
+        NetInfo *net = port_pair.second.net;
+        if (net == nullptr) {
+            continue;
+        }
+
+        // This net doesn't have a driver, probably not valid?
```

Comment:
Undriven nets aren't necessarily a problem (for example if you connect a net to an input but don't drive it); in other arches we tend to skip over them.

Reply:
I believe you mean to an output, but don't drive it.  A disconnected cell primitive input is a bug in most cases I can think of.  It must be either 1/0/signal, rather than "unknown".